hermes/bin/src/runtime_extensions/utils/mul_add.rs

```diff
@@ -0,0 +1,97 @@
+//! Signed Integer Multiply and
+
+use catalyst_types::conversion::from_saturating;
+
+/// A trait providing a saturating multiply-accumulate:
+///
+/// `self + (y * z)` with saturating arithmetic, updating `Self`.
+///
+/// `self` is the Accumulator, and can be any integer but will saturate at max i128 if
+/// its a `u128` or larger.  Otherwise it will saturate within the bounds of its type.
+///
+/// `y` and `z` can be any integer type.
+///
+/// Works across any combination of integer types (signed or unsigned).
+pub trait SaturatingMulAdd<U, V> {
+    /// Multiply and Accumulate Integers.
+    fn mul_add(
+        &mut self,
+        y: U,
+        z: V,
+    );
+}
+
+/// Implement the trait for various integer types.
+macro_rules! impl_saturating_mul_add {
+    ($($t:ty),*) => {
+        $(
+            impl<U, V> SaturatingMulAdd<U, V> for $t
+            where
+                U: Copy
+                    + TryInto<i128>
+                    + std::ops::Sub<Output = U>
+                    + std::cmp::PartialOrd<U>
+                    + num_traits::identities::Zero,
+                V: Copy
+                    + TryInto<i128>
+                    + std::ops::Sub<Output = V>
+                    + std::cmp::PartialOrd<V>
+                    + num_traits::identities::Zero,
+            {
+                fn mul_add(&mut self, y: U, z: V) {
+                    let self128: i128 = from_saturating(*self);
```

Comment:
Even if we work with `u8`, we'll first convert it to `i128`. That's a performance hit, which is probably acceptable in our future use-case for this code, but worth being noticed.